src/jit/morph.cpp

```diff
@@ -13983,39 +13983,62 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
     if (fgGlobalMorph && (oper == GT_ADD) && !tree->gtOverflow() && (op1->gtOper == GT_ADD) && !op1->gtOverflow() &&
         varTypeIsIntegralOrI(typ))
     {
+        GenTree* ad1 = op1->gtOp.gtOp1;
         GenTree* ad2 = op1->gtOp.gtOp2;
 
         if (op2->OperIsConst() == 0 && ad2->OperIsConst() != 0)
         {
-            // This takes
-            //       + (tree)
-            //      / \.
-            //     /   \.
-            //    /     \.
-            //   + (op1) op2
-            //  / \.
-            //     \.
-            //     ad2
+            //  This takes
```

Comment:
Since these methods return bool can you get rid of the `==0` and `!=0` ?

Reply:
Done.

Reply:
Yes, the "((x+icon1)+(y+icon2)) to ((x+y)+(icon1+icon2))" transformation was also unsafe. I added a test for that and fixed the code. The codegen for `Run2`:
before
```
       4803D1               add      rdx, rcx
       4A8D4402F0           lea      rax, bword ptr [rdx+r8-16]
       C3                   ret
```
after
```
       4883C1F8             add      rcx, -8
       4883C2F8             add      rdx, -8
       488D0411             lea      rax, bword ptr [rcx+rdx]
       4903C0               add      rax, r8
       C3                   ret
```